rust/kernel/thread.rs

```diff
@@ -0,0 +1,198 @@
+
+        if result.is_err() {
+            // Creation fails, we need to get back the double boxed closure.
+            //
+            // SAFETY:
+            //
+            // `db` is a proper pointer generated by a `Box::into_raw()` from a box created by us,
+            // if the thread creation fails, no one will consume that pointer.
+            unsafe {
+                Box::from_raw(db);
+            }
+        } else {
+            // Increase the refcount of the task, so that it won't go away if it `do_exit`.
+            //
+            // SAFETY:
+            //
+            // `task` is a valid pointer to newly created thread
+            unsafe {
+                rust_helper_get_task_struct(task);
+            }
+        }
+
+        Ok(Thread { task: result? })
+    }
+
+    /// Wake up the kernel thread, note that a newly created kthread will not run until `wake_up`
+    /// is called.
+    ///
+    /// This function might sleep, don't call in atomic contexts.
+    pub fn wake_up(&self) {
+        // SAFETY:
+        //
+        // `task` is a valid pointer to a kernel thread structure, and it's not stopped because
+        // `stop` will consume the [`Thread`].
+        unsafe {
+            bindings::wake_up_process(self.task);
+        }
+    }
+
+    /// Stop the the kernel thread:
+    ///
+    /// * If the thread hasn't been waken up after creation, the thread closure won't be called,
+    /// and will return `EINTR`. (Note that a thread may not be waken up even after `wake_up`
+    /// is called).
+    /// * Otherwise, wait for the closure to return or the thread `do_exit` itself.
+    ///
+    /// Consume the [`Thread`] so that it's not accessible. Return the result of the thread
+    /// closure (or the exit code in KernelResult format).
+    ///
+    /// This function might sleep, don't call in atomic contexts.
+    pub fn stop(self) -> KernelResult<()> {
+        let ret;
+        // SAFETY:
+        //
+        // `task` is a valid pointer to a kernel thread structure, and it's not stopped because
+        // `stop` will consume the [`Thread`].
+        unsafe { ret = bindings::kthread_stop(self.task) }
+
+        if ret == 0 {
+            Ok(())
+        } else {
+            Err(Error::from_kernel_errno(ret))
+        }
+    }
+}
+
+impl Drop for Thread {
+    fn drop(&mut self) {
+        // Decrease the refcount of the thread, the actuall thread may still be running after we
+        // `drop` the [`Thread`].
+        //
+        // SAFETY:
+        //
+        // At least one refcount is held by `Thread::new` and refcount of [`task_struct`] is
+        // implemented by atomics.
+        unsafe {
+            rust_helper_put_task_struct(self.task);
+        }
+    }
+}
+
+/// Try to give up the cpu and let another thread to run, mapping to kernel's `schedule` function.
+/// A similar concept to `std::thread::yield_now`.
+///
+/// This function might sleep, don't call in atomic contexts.
+pub fn schedule() {
```

Comment:
Should we call this `yield_now` for consistency with rust or stick with `schedule`?

@ojeda @alex -- do you have an opinion here?

Reply:
In general, I'd say it is best to use kernel terminology unless the type matches perfectly the Rust concept to the point we can use the same API as the Rust type.

The main reason is that most developers judging and reading this will be kernel ones, and it is the gap we need to cross (at least meanwhile we are not in mainline).

In this case, since the type has a different API and it is not a userspace thread anyway, I'd favor prefer the kernel terminology.

Note that we should, nevertheless, add doc aliases wherever appropriate to make it easier for newcomers from Rust to find out the equivalents, i.e. here to `yield_now`.

Reply:
I like this rationale.